### ISSUE TITLE
Intermittent error with affine transforms in 0.4.0

### ISSUE THREAD

**Issue by jni** [napari/napari#1880]
## 🐛 Bug

@DragaDoncila and I aren't sure if this is fixed in master, but wanted to write it down in case it's still around and an obvious fix to @sofroniewn. 

When adding layers to napari, we sometimes see this error:

```pytb
ERROR:root:Unhandled exception:
Traceback (most recent call last):
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/_qt/qt_viewer.py", line 758, in dropEvent
    self.viewer.open(filenames, stack=bool(shift_down))
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/components/add_layers_mixin.py", line 341, in open
    self._add_layers_with_plugins(
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/components/add_layers_mixin.py", line 409, in _add_layers_with_plugins
    new = self._add_layer_from_data(*_data)
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/components/add_layers_mixin.py", line 481, in _add_layer_from_data
    layer = add_method(data, **(meta or {}))
  File "<string>", line 3, in add_tracks
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/components/add_layers_mixin.py", line 73, in add_layer
    self._update_layers(layers=[layer])
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/components/viewer_model.py", line 313, in _update_layers
    layer._slice_dims(
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/layers/base/base.py", line 846, in _slice_dims
    self._update_dims()
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/layers/base/base.py", line 518, in _update_dims
    self.refresh()
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/layers/base/base.py", line 893, in refresh
    self._update_value_and_status()
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/layers/base/base.py", line 904, in _update_value_and_status
    self._value = self.get_value()
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/layers/base/base.py", line 866, in get_value
    return self._get_value()
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/layers/tracks/tracks.py", line 242, in _get_value
    coords = np.array(self.coordinates)
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/layers/base/base.py", line 900, in coordinates
    return tuple(self._transforms[1:].simplified.inverse(self.position))
  File "/home/draga/anaconda3/envs/napari/lib/python3.8/site-packages/napari/utils/transforms/transforms.py", line 328, in __call__
    linear_matrix[
ValueError: could not broadcast input array from shape (4,4) into shape (2,2)
```

Again, this happens intermittently so my guess is that it is a race condition with the dimensionality of the viewer before and after the layers are added.

## Environment

This is on napari 0.4.0

**Comment by sofroniewn** [comment]
> When adding layers to napari, we sometimes see this error:

Can you provide a bit more detail or an example? Ideally not using tracks but just an / some image layers? Looks like mixing 2D and 4D? Just reading the trace and without knowing the plugin or data, I can't tell much about what's going on ...

**Comment by jni** [comment]
Yeah we'll get there, we are just under a deadline. =) But @DragaDoncila reports that this happened also with her satellite (2D+t images) data.

**Comment by sofroniewn** [comment]
From speaking with @jni, sounds like this is drag and drop specific. I tried dragging and dropping a bunch of 2d / 3d tiff stack etc data, including large datasets and I couldn't reproduce. Happy to follow up when more detail provided

**Comment by DragaDoncila** [comment]
I've been able to semi consistently produce this using nd2 images and ome-zarr images (being opened using their respective plugins ome_zarr and nd2-dask) by:
- opening napari from command line and drag&drop
- running a python script which opens the file
- running a python console, opening a viewer and drag&dropping a file

I've never been able to produce this in an ipython session for some reason, and it seems to happen more often when the viewer is freshly loaded*.

The smallest file I can provide that can consistently (but still intermittently) produce this error is [here](https://drive.google.com/file/d/198NrM7mPejsi5Ps2FHfqSTDjvd4CXnXG/view?usp=sharing)

*freshly loaded - hasn't opened any other files beforehand

**Comment by DragaDoncila** [comment]
I should also mention that the viewer still opens the file (or at least some of it), even if the error occurs.

With the nd2 files, one layer is opened, and the remaining three are not.

With the label example I provided above there's only one layer - the error occurs but the layer is still opened, and I'm able to zoom in to lower levels of the pyramid with no issue

**Comment by adamltyson** [comment]
I'm also seeing this, and it usually happens when loading a tiff series (either with `napari-ndtiffs` or `magic_imread`). It has happened (but not every time) with every method of loading I've tried.

I noticed it in `0.4.0` and it persists in `0.4.1rc2`.

**Comment by camiel-m** [comment]
Hi, I'm curious if there are any updates on this. I'm having the same issue with 2D+t data (tif files) and would like to know if there's some trick to circumvent it in the current version of napari.

**Comment by sofroniewn** [comment]
Hi @camiel-m ah no unfortunately no update here, sorry it seems to have fallen through the cracks a little. I've just bumped the priority status on this one.

I wasn't able to reproduce myself which made debugging hard. Can you maybe provide some details about that tiff file. Ideally it there was a script to make such a file from a numpy array then i could try with that - maybe the array is 100 planes with each place 1024 x 1024 say, or other info like that.

**Comment by camiel-m** [comment]
Hi Nicholas,

That’s great to hear and I just want to say I really enjoy napari and how actively it’s being supported/developed.
On to the issue, my data consists of 376 files that are 512x512 pixels. The files are saved from numpy arrays with dtype int16.

Camiel

**Comment by andy-sweet** [comment]
I was thinking about picking this issue up, but after some initial poking around the code, I suspect it is not possible to reproduce anymore for the following reasons.

1. The crash occurred because the value returned by Layer::coordinates had a lower dimensionality than that of the layer's transform.
2. [Layer::world_to_data](https://github.com/napari/napari/blob/07df2af8282265dbbdbb1fb0f8aeb3c218150f94/napari/layers/base/base.py#L1006) was added since the last comment on this bug, and ensures that the dimensionality of the input coordinates is the same as the dimensionality of the layer (and thus I assume the dimensionality of all its transform).

Maybe more importantly, Layer::coordinates has been deprecated and does not seem to be used within the napari code base anymore.

However, I'm still unsure of why there was a dimension mismatch, and if padding with zeros is really what we want to do. It would be great if someone who was able to reproduce this bug could try again with the latest version of napari to see if there are any related issues.

**Comment by sofroniewn** [comment]
Thanks for looking into this @andy-sweet, we can leave it open for a little more in case someone runs into similar problem on latest version of napari (say > `0.4.8`) but agreed, it might no longer be so relevant due to underlying code changes. If we don't hear anything I'll close issue, but people should feel free to reopen/ comment more if they run into additional problems

**Comment by DragaDoncila** [comment]
Hi @andy-sweet fwiw I used to see this semi regularly and have not seen it in a while. Recently (last three weeks) I've worked with a variety of tiffs (2D+t), ome-zarrs (2D+t) and nd2 files (3D+t) recording demos for a talk and haven't run into this again.

Obviously not conclusive proof, but given your investigations it's likely this is no longer a bug.

I'll keep an explicit eye out for it in the next few days. I was never able to reproduce consistently.
